### DISCUSSION TITLE
RTR_ListFiles returns empty "resources"

### DISCUSSION THREAD

**Discussion by TaihouKai** [CrowdStrike/falconpy#880] [Q&A] [answered]
I tried to `get` files with the following function:
```
def cmd_rtr(rtr_admin, session_id, base_cmd, cmd, timesleep, timemax):
    # Execute RTR command
    cloud_request_id = exec_cmd_rtr(rtr_admin, session_id, base_cmd, cmd)["cloud_request_id"]
    print(f"Command executed: {cloud_request_id}")

    # Wait for RTR command to finish
    res = None
    completed = False
    count = 0
    while (not completed):
        time.sleep(timesleep)
        res = check_cmd_rtr(rtr_admin, cloud_request_id)
        completed = res["cmd_completed"]
        if (count * timesleep > timemax):
            print(f"Timeout! Please try again.")
            sys.exit()

    # RTR command stdout
    res_out = res["cmd_status"]["body"]["resources"][0]

    return {"std_out": res_out["stdout"], "std_err": res_out["stderr"], "cloud_request_id": cloud_request_id}
```

And use `RTR_ListFiles` to list all files after `get`:
```
files = rtr.RTR_ListFiles(session_id=session_id)["body"]["resources"]
```

But, sometimes (not always) it returns:
```
{
  'status_code': 200, 
  'headers': {
    'Server': 'nginx', 
    'Date': 'Tue, 17 Jan 2023 02:14:39 GMT', 
    'Content-Type': 'application/json', 
    'Content-Length': '158', 
    'Connection': 'keep-alive', 
    'Content-Encoding': 'gzip', 
    'X-Cs-Region': 'us-1', 
    'X-Cs-Traceid': '01e03607-8ae2-4080-88a6-cc5e427abcc7', 
    'X-Ratelimit-Limit': '6000', 
    'X-Ratelimit-Remaining': '5988', 
    'Strict-Transport-Security': 'max-age=31536000; includeSubDomains'
  }, 
  'body': {
    'meta': {
      'query_time': 0.007406469, 
      'powered_by': 'empower-api', 
      'trace_id': '01e03607-8ae2-4080-88a6-cc5e427abcc7'
    }, 
  'resources': [], 
  'errors': []
  }
}
```

Is it because the file is not ready yet? I thought I've waited for the `GET` command to complete in `cmd_rtr` function.

If this is the case, how should I tell the program to wait until the file is ready, besides a loop check until it finds something?

**Comment by jshcodes** [comment, marked as answer]
Hi @TaihouKai -

Thanks for the question!

You're correct, this is most likely occurring because the file transfer has not yet completed.  You're also on the correct path to address this. Similar to how you check for command execution with `cloud_request_id` using the [__`RTR_CheckCommandStatus`__](https://www.falconpy.io/Service-Collections/Real-Time-Response.html#rtr_checkcommandstatus) operation above, we can implement a loop check.  (You could add a sleep timer for a few seconds, or minutes, depending on your scenario.)

> __Regarding the command execution loop check__: You can check for the value of complete in the results returned from [__`RTR_CheckCommandStatus`__](https://www.falconpy.io/Service-Collections/Real-Time-Response.html#rtr_checkcommandstatus), but it is safer to check the contents of `stdout` and `stderr`.  While this happens rarely, when the result exceeds the available size for `stdout`, it is chunked across multiple pages (using the `sequence_id` value). When this occurs, the value of `complete` will be __False__.

**Reply by TaihouKai** [reply]
Thank you so much for your answer!
